_posts/2022-07-22-polymorphic-table-functions.md

```diff
@@ -0,0 +1,192 @@
+---
+layout: post
+title: "Diving into polymorphic table functions with Trino"
+author: "Kasia Findeisen, Brian Olsen, and Cole Bowden"
+excerpt_separator: <!--more-->
+---
+
+In the Trino community, we know that being the coolest query engine is a tough
+job. We boldly face the intricacies of the SQL standard to bring you the newest
+and most powerful features. Today, we proudly announce that as of release 381,
+Trino is on its way to full support for polymorphic table functions (PTFs).
+
+In this blog post, we are explaining the concept of table functions and 
+exploring how they can be leveraged. We also look at what we have already 
+implemented, and take a sneak peek into the future.
+
+<!--more-->
+
+### Definition time
+
+There are several kinds of functions you can call in a SQL query: scalar
+functions, aggregate functions, and window functions. They might process the
+input row by row (scalar) or all at once (aggregate). One thing they have in
+common is that they return scalar values. Table functions are different. They
+return tables. In a query, they can appear in any place where a table reference
+shows up such as a `FROM` clause:
+
+```sql
+SELECT
+  *
+FROM
+  TABLE(my_table_function('foo'));
+```
+
+You can also use table functions in joins:
+
+```sql
+SELECT
+  *
+FROM
+  TABLE(my_table_function('bar'))
+JOIN
+  TABLE(another_table_function(1, 2, 3))
+ON true;
+```
+
+Polymorphic table functions (PTFs) are a subset of table functions where the
+schema of the returned table is determined dynamically. The returned table
+schema can depend on the arguments you pass to the function.
+
+### OK, but why are we so excited?
+
+We are excited because this feature is a real game changer! Polymorphic table
+functions make SQL extensible, provide a framework for processing data in
+previously impossible ways, and can act as a bridge between the Trino engine and
+external systems or resources you might need for processing your data.
+Additionally, polymorphic table functions are standard SQL, and they are very
+convenient to use.
+
+### What is available in Trino today?
+
+So far, we have added a framework for table functions which can be executed by
+the connector. Although this is not the full PTF feature yet, we couldn't wait
+to bring it to life. We added query pass-through table functions for JDBC-based
+connectors and ElasticSearch. They mostly go by the name `query`, and they take
+a single argument, that being the query text:
+
+```sql
+SELECT
+  *
+FROM
+  TABLE(
+    postgresql.system.query(query =>
+        'SELECT
+          name
+        FROM
+          tpch.nation
+        WHERE
+          nationkey = 0'
+    )
+  );
+```
+
+And this will return:
+
+```sql
+  name
+---------
+ ALGERIA
+(1 row)
+```
+
+Something you can’t notice from that example is that when you’re passing that
+“query” argument, it’s pushing the entire query down to PostgreSQL to process
```

Comment:
“Pushing” is misleading. That’s usually used in the context of the capabilities of the optimizer. In the case of the query function, it literally executes the query in the remote system without any involvement from the optimizer.